Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,459 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// This is a companion to System.swift that provides only Linux specials: either things that exist
+// only on Linux, or things that have Linux-specific extensions.
+
+#if os(Linux)
+
+import CNIOLinux
+
+// we stuff the event type into the user data for the sqe together with
+// the fd to match the events without needing any memory allocations or
+// references. Just shift in the event type in the upper 32 bits.
+
+internal enum CqeEventType : Int {
+    case poll = 1, pollModify, pollDelete // start with 1 to not get zero bit patterns for stdin
+}
+
+internal enum UringError: Error {
+    case loadFailure
+    case uringSetupFailure
+    case uringWaitCqeFailure
+    case uringWaitCqeTimeoutFailure
+}
+
+internal extension TimeAmount {
+    func kernelTimespec() -> __kernel_timespec {
+        var ts = __kernel_timespec()
+        ts.tv_sec = self.nanoseconds / 1_000_000_000
+        ts.tv_nsec = self.nanoseconds % 1_000_000_000
+        return ts
+    }
+}
+
+public struct UringEvent {
+    var fd : Int32
+    var pollMask : UInt32
+}
+
+public class Uring {
+    public static let POLLIN: CUnsignedInt = numericCast(CNIOLinux.POLLIN)
+    public static let POLLOUT: CUnsignedInt = numericCast(CNIOLinux.POLLOUT)
+    public static let POLLERR: CUnsignedInt = numericCast(CNIOLinux.POLLERR)
+    public static let POLLRDHUP: CUnsignedInt = numericCast(CNIOLinux.EPOLLRDHUP.rawValue) // FIXME: - POLLRDHUP not in ubuntu headers?!
+    public static let POLLHUP: CUnsignedInt = numericCast(CNIOLinux.POLLHUP)
+
+    private var ring = io_uring()
+    // FIXME: These should be tunable somewhere, somehow. Maybe environment vars are ok, need to discuss with SwiftNIO team.
+    private let ringEntries: CUnsignedInt = 8192 // this is a very large number due to some of the test that does 1K registration mods
+    private let cqeMaxCount : UInt32 = 4096 // shouldn't be more than ringEntries, this is the max chunk of CQE we take.
+        
+    var cqes : UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>
+    var fdEvents = [Int32: UInt32]() // fd : event_poll_return
+    var emptyCqe = io_uring_cqe()
+
+    // FIXME: This is not thread safe, needs some other mechanism for guaranteeing single io_uring_load
```

Comment:
The way to guarantee a single load is to put this inside a `static let`:

```swift
private static let initializedUring: Bool = {
    CNIOLinux.CNIOLinux_io_uring_load() != 0
}()
```

`static let`s of this form are guaranteed to be initialised as a `dispatch_once`, which is exactly equivalent to `pthread_once` for our use-case.

Reply:
Thanks, exactly what I was looking for (although should be ==0 for the expression when moved there :-). Fixed.